File: extract_faq.pl
#!/usr/bin/perl
use strict;
use warnings;
use Cwd qw(abs_path);
use File::Basename qw(dirname);

# --- Add local lib path ---
my $script_dir = dirname(abs_path($0));
my $lib_path = "$script_dir/JSON-4.10/lib";
unshift @INC, $lib_path;

use JSON;

# --- Configuration ---
my $units_file = 'units.json';
my $faq_file = 'faq.json';
my $debug_html_file = 'debug_unit.html'; # File to save HTML for debugging

# --- Main Logic ---
sub main {
    my $json_text = read_file($units_file);
    my $units = decode_json($json_text);
    my %faq_data;

    print "Starting FAQ extraction for a single unit for debugging...\n";

    my $unit = $units->[0]; # Just get the first unit for debugging
    my $id = $unit->{id};
    my $href = $unit->{href};

    print "Fetching FAQ for unit $id: $unit->{name}\n";

    my $html = `curl -sL "$href"`;
    if ($html) {
        write_file($debug_html_file, $html);
        print "  -> Saved HTML to $debug_html_file\n";
        my $faq_text = extract_faq_from_html($html);
        if ($faq_text) {
            $faq_data{$id} = $faq_text;
            print "  -> Found FAQ.\n";
        } else {
            print "  -> No FAQ found.\n";
        }
    } else {
        print "  -> Failed to fetch URL.\n";
    }

    # write_file($faq_file, encode_json(\%faq_data));
    # print "FAQ extraction complete. Data saved to $faq_file\n";
}

# --- Subroutines ---
sub read_file {
    my ($filename) = @_;
    open(my $fh, '<:encoding(UTF-8)', $filename) or die "Could not open file '$filename' $!";
    my $content = do { local $/; <$fh> };
    close($fh);
    return $content;
}

sub write_file {
    my ($filename, $content) = @_;
    open(my $fh, '>:encoding(UTF-8)', $filename) or die "Could not open file '$filename' $!";
    print $fh $content;
    close($fh);
}

sub extract_faq_from_html {
    my ($html) = @_;
    # The regex might be the issue. Let's try a broader one first.
    if ($html =~ /Special\s+Abilities\s+Errata\s*&\s*FAQ/i) {
        my ($faq_section) = $html =~ /(<a name="specials">.*?)(<a name=|<div id="footer">)/si;
        if ($faq_section) {
            $faq_section =~ s/<[^>]*>//g; # Strip HTML tags
            $faq_section =~ s/^\s+|\s+$//g; # Trim whitespace
            return $faq_section;
        }
    }
    return '';
}

# --- Run ---
main();

File: extract_faq.py
import json
import requests
import time
from bs4 import BeautifulSoup

UNITS_FILE = 'units.json'
FAQ_FILE = 'faq.json'

def main():
    with open(UNITS_FILE, 'r', encoding='utf-8') as f:
        units = json.load(f)

    faq_data = {}
    print(f"Starting FAQ extraction for {len(units)} units...")

    for i, unit in enumerate(units):
        unit_id = unit.get('id')
        href = unit.get('href')

        if not href:
            print(f"Skipping unit {unit_id} due to missing href.")
            continue

        print(f"Fetching FAQ for unit {i+1}/{len(units)}: {unit.get('name')}")

        try:
            response = requests.get(href, timeout=10)
            response.raise_for_status()  # Will raise an HTTPError for bad responses
            html = response.text

            soup = BeautifulSoup(html, 'html.parser')

            faq_header = soup.find(lambda tag: "Special Abilities Errata & FAQ" in tag.get_text())

            if faq_header:
                faq_content = ""
                for sibling in faq_header.find_next_siblings():
                    if sibling.name == 'a' and sibling.get('name') == 'specials':
                        continue
                    if (sibling.name == 'a' and sibling.get('name')) or (sibling.name == 'div' and sibling.get('id') == 'footer'):
                        break
                    faq_content += sibling.get_text(separator='\\n', strip=True) + '\\n'

                if faq_content.strip():
                    faq_data[unit_id] = faq_content.strip()
                    print("  -> Found FAQ.")
                else:
                    print("  -> FAQ section found, but it's empty.")
            else:
                print("  -> No FAQ section found.")

        except requests.exceptions.RequestException as e:
            print(f"  -> Failed to fetch URL: {e}")

        time.sleep(0.5)  # Be a good netizen

    with open(FAQ_FILE, 'w', encoding='utf-8') as f:
        json.dump(faq_data, f, indent=2)

    print(f"FAQ extraction complete. Data saved to {FAQ_FILE}")

if __name__ == '__main__':
    main()
